docs: Kubernetes Kit session replication (#1867)
vaadin/docs: articles/tools/azure/session-replication.adoc

---
title: Session Replication
description: Best practices to achieve session replication
order: 100
---

= Session Replication

For a Vaadin application to reach the goals of High Availability and Down Scaling, it is fundamental that the HTTP session is correctly distributed between multiple servers.

If the server processing user requests becomes unavailable, it should be transparently replaced by another one, in a way that the user should be able to continue using the application without any disruption.

> The user may notice a temporarily offline notification or the loading indicator spawn, but browser page is not reload and his work can continue without losing any data.

Reviewer: ```suggestion
The user may notice a temporarily offline notification or the loading indicator spawn, but the browser page is not reloaded and work can continue without losing any data.
```

> Speaking about HTTP session, this means that attributes must be constantly stored in a shared/distributed storage, so that newly created HTTP sessions may filled with the most recent user data.

Reviewer: ```suggestion
This means that attributes must be constantly stored in a shared/distributed storage, so that newly created HTTP sessions may filled with the most recent user data.
```

> == Vaadin Kubernetes Kit Session Replication

Reviewer: ```suggestion
== Kubernetes Kit Session Replication
```

> Vaadin Kubernetes Kit helps with session replication, by transfer HTTP session data on a backed session storage (Hazelcast or Redis) at every request.

Reviewer: ```suggestion
Kubernetes Kit helps with session replication, by transferring HTTP session data via a backend session storage (Hazelcast or Redis) at every request.
```

> When a server becomes unavailable, the one taking its place loads user data from the session storage and fills the new server side HTTP session object.

Reviewer: ```suggestion
When a server becomes unavailable, the one taking its place loads user data from the session storage and fills the new server-side HTTP session object.
```

To transfer information to the session storage, the HTTP session attributes are encoded in a binary format, using https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/index.html[Java serialization].

> This means that to have session replication working, the application code should be written in way that all object stored into HTTP session are eligible for Java serialization, meaning they must implement a special marker interface, [classname]`java.io.Serializable`.

Reviewer: ```suggestion
This means that to have session replication working, the application code should be written in way that all objects stored in the HTTP session are eligible for Java serialization, meaning they must implement a special marker interface, [classname]`java.io.Serializable`.
```

> Objects that for any reasons cannot be serializable (e.g. third party classes) must be defined `transient`, so they are skipped by the serialization process.

Reviewer: ```suggestion
Objects that for any reason cannot be serializable (e.g. third party classes) must be defined `transient`, so they are skipped by the serialization process.
```

Of course, after deserialization transient fields are null.

> For Spring application, Vaadin Kubernetes Kit is also capable to overcome this limitation.

Reviewer: ```suggestion
For Spring applications, Kubernetes Kit is also able to overcome this limitation.
```

> Transient fields referencing Spring beans are detected during serialization and metadata is stored along the session attributes.

Reviewer: ```suggestion
Transient fields referencing Spring beans are detected during serialization and metadata is stored along with the session attributes.
```

On deserialization, transient fields matching metadata are injected with the corresponding Spring bean.

== Vaadin and HTTP Session

> A Vaadin application is composed by UI objects that acts as root nodes of graphs of components composition that represents the web page the user interact with in through the browser.

Reviewer: ```suggestion
A Vaadin application is comprised of UI objects that act as root nodes of graphs of components that represent the web page the user interacts with in through the browser.
```

> UIs are collected into [classname]`VaadinSession` that is stored in the HTTP session, to preserve the server side state across request.

Reviewer: ```suggestion
UIs are collected into a [classname]`VaadinSession` that is stored in the HTTP session, to preserve the server-side state across requests.
```

> This means that to have session replication working, the application code should be written in way that all object stored into HTTP session are eligible for Java serialization, meaning they must implement a special marker interface, [classname]`java.io.Serializable`.

Reviewer: Repeated segment from line 25.

Author: I tried to rephrase the sentence

For a deeper insight of serialization process, consult [Java documentation] website.

== Vaadin Session Replication Tips

The aim of the following sections is to give tips and best practices on how to code a fully serializable Vaadin applications.

=== Use Serializable Object as UI Components Members

> When writing a component make sure that all its non transient members implements [classname]`Serializable`.

Reviewer: ```suggestion
When writing a component make sure that all its non-transient members implement [classname]`Serializable`.
```

> All Vaadin UI components and listeners are already marked as serializable, so if extending or implementing them it is not necessary to explicitly add the `Serializable` implementation statement.

Reviewer: ```suggestion
All Vaadin UI components and listeners are already marked as serializable, so if a class extends or implements one of them, it is not necessary to explicitly add the `Serializable` implementation statement.
```

However, it is mandatory to do that on non UI components that may be referenced in the UI graph.

> For example, if a view is holding a reference to a DTO or a configuration object, it must be marked as serializable, and all its member must be serializable.

Reviewer: ```suggestion
For example, if a view is holding a reference to a data transfer object or a configuration object, it must be marked as serializable, and all its members must be serializable.
```

[source,java]
----
class ContactForm extends VerticalLayout { // (1)

>     private Product product; // (2)

Reviewer: I don't think the label is meant to be here.

Author: I put the label in two places because the related description covers both the lines of code
- Product instance is referenced by ContactForm
- it must be Serializable

Can this be expressed better using different labels?

Reviewer 2: I made a suggestion to reorganize the callouts.

}

class Product implements Serializable { // (2)

>     private Category category; // (3)

Reviewer: I don't think the label is meant to be here.

Author: I put the label in two places because the related description covers both the lines of code
- Category is referenced by Product
- it also must be Serializable

Can this be expressed better using different labels?

Reviewer 2: I made a suggestion to reorganize the callouts.

> class ContactForm extends VerticalLayout { // (1)
> 
>     private Product product; // (2)
> 
> }
> 
> class Product implements Serializable { // (2)
> 
>     private Category category; // (3)
> }
> 
> class Category implements Serializable { // (3)
> 
> }
> ----
> 
> 1. The [classname]`ContactForm` class extends a Vaadin component that is already serializable.
> 1. [classname]`Product` instance is referenced by [classname]`ContactForm`, so it must be [classname]`Serializable`.
> 1. [classname]`Category` is referenced by [classname]`Product`, so it also must be [classname]`Serializable`.

Reviewer 2: ```suggestion
class ContactForm extends VerticalLayout { // (1)

    private Product product; // (2)

}

class Product implements Serializable { // (3)

    private Category category; // (4)
}

class Category implements Serializable { // (5)

}
----

1. The [classname]`ContactForm` class extends a Vaadin component that is already serializable.
2. The [classname]`ContactForm` class has a field of type [classname]`Product`.
Therefore, the [classname]`Product` class must also be made serializable.
3. The [classname]`Product` class is made serializable by extending [interfacename]`Serializable`.
4. The [classname]`Product` class has a field of type [classname]`Category`.
Therefore, the [classname]`Category` class must also be made serializable.
5. The [classname]`Category` class is made serializable by extending [interfacename]`Serializable`.
```

> === Transient Fields for Not Serializable Objects

Reviewer 2: ```suggestion
=== Transient Fields for Non-Serializable Objects
```

        );
    }

    private OrderService getOrderService() {
        return service;
    }
}
----

Another way, to avoid adding the transient field to the main class, is to reference the non serializable object in a serializable proxy, that exposes only the required methods.

[source,java]
----
class OrderEditor {

    private ComboBox<OrderState> status;
    
    OrderEditor(OrderService service) {
        this.service = service;
        OrderStateLabelGeneratorProxy proxy = new OrderStateLabelGeneratorProxy(service);
        status.setItemLabelGenerator(item -> 
            proxy.humanReadableState(item)
        );
    }

    private static class OrderStateLabelGeneratorProxy
        implements Serializable {
        
        private final transient OrderService service;

        OrderStateLabelGeneratorProxy(OrderService service) {
            this.service = service;
        }

        String humanReadableState(OrderState state) {
            return service.humanReadableState(item);
        }
    }
}
----

== Session Replication Issues

Despite from applying mentioned tips, still session replication may fail because of issues during serialization or deserialization.

> When facing exceptions on serialization phase, it may be useful set the `sun.io.serialization.extendedDebugInfo` system property to make Java serialization verbose.

Reviewer 3: Should we make this an AsciiDoc `TIP` to make it stand out?

A common Vaadin application extensively uses lambda expression for components listeners, binder, etc.

When serializing and deserializing lambda expressions, it may happen to face [classname]`ClassCastException` with cryptic messages, such as `SerializedLambda cannot be cast to class <className>`  (on serialization) or `SerializedLambda cannot be assigned to field <fieldName> of type <className>` (on deserialization).

> Usually the cause is a `self reference`, meaning that the lambda expression captures an object instance, but the expression is itself a member of the object graph of the captured object.

Reviewer 4: Would this be a better formatting? I wonder if monospace is the best one to use.
```suggestion
Usually the cause is a "self reference", meaning that the lambda expression captures an object instance, but the expression is itself a member of the object graph of the captured object.
```
